src/util/ExpressionUtils.cpp

```diff
@@ -106,6 +106,38 @@ ExpressionUtils::pullOrsImpl(LogicalExpression *expr,
     }
 }
 
+Expression *ExpressionUtils::flattenInnerLogicalAndExpr(const Expression *expr) {
+    auto matcher = [](const Expression *e) -> bool {
+        return e->kind() == Expression::Kind::kLogicalAnd;
+    };
+    auto rewriter = [](const Expression *e) -> Expression * {
+        pullAnds(const_cast<Expression *>(e));
+        return e->clone().release();
+    };
+
+    return RewriteVisitor::transform(expr, std::move(matcher), std::move(rewriter));
+}
+
+Expression *ExpressionUtils::flattenInnerLogicalOrExpr(const Expression *expr) {
+    auto matcher = [](const Expression *e) -> bool {
+        return e->kind() == Expression::Kind::kLogicalOr;
+    };
+    auto rewriter = [](const Expression *e) -> Expression * {
+        pullOrs(const_cast<Expression *>(e));
+        return e->clone().release();
+    };
+
+    return RewriteVisitor::transform(expr, std::move(matcher), std::move(rewriter));
+}
+
+Expression *ExpressionUtils::flattenInnerLogicalExpr(const Expression *expr) {
+    auto* andFlattenExpr = flattenInnerLogicalAndExpr(expr);
+    auto* allFlattenExpr = flattenInnerLogicalOrExpr(andFlattenExpr);
+    delete(andFlattenExpr);
```

Comment:
Use std::unique_ptr<Expression> to replace `delete`

Reply:
ACK.